src/core/or/circuitpadding.c

```diff
@@ -0,0 +1,2370 @@
+      histogram_total += histogram[b];
+  } else if (state->iat_dist.type == CIRCPAD_DIST_NONE) {
+    /* We have a histogram, but it's immutable */
+    histogram = state->histogram;
+    histogram_total = state->histogram_total;
+  } else {
+    /* Sample from a fixed IAT distribution and return */
+    double val = circpad_distribution_sample(state->iat_dist);
+    /* These comparisons are safe, because the output is in the range
+     * [0, 2**46], and double has a precision of 53 bits. */
+    val = MAX(0, val);
+    val = MIN(val, state->range_sec*TOR_USEC_PER_SEC);
+
+    /* This addition is exact: val is at most 2**16 * USEC_PER_SEC ~= 2**46,
+     * start_usec is at most 2**32, and doubles have a precision of 53 bits. */
+    val += start_usec;
+    /* Clamp the distribution at infinite delay val */
+    return (circpad_delay_t)MIN(tor_llround(val), CIRCPAD_DELAY_INFINITE);
+  }
+
+  bin_choice = crypto_rand_int(histogram_total);
+
+  /* Skip all the initial zero bins */
+  while (!histogram[i]) {
+    i++;
+  }
+  curr_weight = histogram[i];
+
+  // TODO: This is not constant-time. Pretty sure we don't
+  // really need it to be, though.
+  while (curr_weight < bin_choice) {
+    i++;
+    tor_assert(i < state->histogram_len);
+    curr_weight += histogram[i];
+  }
+
+  tor_assert(i < state->histogram_len);
+  tor_assert(histogram[i] > 0);
+
+  // Store this index to remove the token upon callback.
+  if (state->token_removal != CIRCPAD_TOKEN_REMOVAL_NONE) {
+    mi->chosen_bin = i;
+  }
+
+  if (i == state->histogram_len-1) {
+    if (state->token_removal != CIRCPAD_TOKEN_REMOVAL_NONE &&
+        mi->histogram[i] > 0) {
+      mi->histogram[i]--;
+    }
+
+    // Infinity: Don't send a padding packet. Wait for a real packet
+    // and then see if our bins are empty or what else we should do.
+    return CIRCPAD_DELAY_INFINITE;
+  }
+
+  tor_assert(i < state->histogram_len - 1);
+
+  bin_start = circpad_histogram_bin_to_usec(mi, i);
+  bin_end = circpad_histogram_bin_to_usec(mi, i+1);
+
+  /* Truncate the high bin in case it's the infinity bin:
+   * Don't actually schedule an "infinite"-1 delay */
+  bin_end = MIN(bin_end, start_usec+state->range_sec*TOR_USEC_PER_SEC*2);
+
+  // Sample uniformly between histogram[i] to histogram[i+1]-1,
+  // but no need to sample if they are the same timeval (aka bin 0 or bin 1).
+  if (bin_end <= bin_start+1)
+    return bin_start;
+  else
+    return (circpad_delay_t)crypto_rand_uint64_range(bin_start, bin_end);
+}
+
+/**
+ * Sample a value from the specified probability distribution.
+ *
+ * This performs inverse transform sampling
+ * (https://en.wikipedia.org/wiki/Inverse_transform_sampling).
+ *
+ * XXX: These formulas were taken verbatim. Need a floating wizard
+ * to check them for catastropic cancellation and other issues (teor?).
+ * Also: is 32bits of double from [0.0,1.0) enough?
+ */
+static double
+circpad_distribution_sample(circpad_distribution_t dist)
```

Comment:
Not sure why we are passing by value here.. I think The Right Thing would be to pass a constified pointer to the dist.

Reply:
Hrrm actually passing these on the stack means that inlining and/or fastcall (register argument passing) can take place (and does on my compiler). So it's actually much faster than pushing a pointer from the stack onto the stack, dereferencing twice, and then copying to registers to do the math. The reason is because this is a tiny struct with three arguments and the compiler figures that out. Not sure re pointer.

Reply:
Sounds good. no need to do this.